aiida/tools/graph/graph_traversers.py

```diff
@@ -0,0 +1,287 @@
+
+    if not isinstance(starting_pks, (list, set, tuple)):
+        raise TypeError('starting_pks must be of type list, set or tuple\ninstead, it is {}'.format(type(starting_pks)))
+
+    if not starting_pks:
+        if get_links:
+            output = {'nodes': set(), 'links': set()}
+        else:
+            output = {'nodes': set(), 'links': None}
+        return output
+
+    operational_set = set()
+    for single_pk in starting_pks:
+        if not isinstance(single_pk, int):
+            raise TypeError(
+                'element provided in starting_pks cannot be pk because it is of type: {}'.format(type(single_pk))
+            )
+        operational_set.add(single_pk)
+
+    query_nodes = orm.QueryBuilder()
+    query_nodes.append(orm.Node, project=['id'], filters={'id': {'in': operational_set}})
+    existing_pks = {pk[0] for pk in query_nodes.all()}
+    missing_pks = operational_set.difference(existing_pks)
+    if missing_pks:
+        raise exceptions.NotExistent(
+            'The following pks are not in the database and must be pruned before this call: {}'.format(missing_pks)
+        )
+
+    rules = []
+    basket = Basket(nodes=operational_set)
+
+    # When max_iterations is finite, the order of traversal may affect the result
+    # (its not the same to first go backwards and then forwards than vice-versa)
+    # In order to make it order-independent, the result of the first operation needs
+    # to be stashed and the second operation must be performed only on the nodes
+    # that were already in the set at the begining of the iteration: this way, both
+    # rules are applied on the same set of nodes and the order doesn't matter.
+    # The way to do this is saving and seting the walkers at the right moments only
+    # when both forwards and backwards rules are present.
+    if links_forward and links_backward:
+        stash = basket.get_template()
+        rules += [RuleSaveWalkers(stash)]
+
+    if links_forward:
+        query_outgoing = orm.QueryBuilder()
+        query_outgoing.append(orm.Node, tag='sources')
+        query_outgoing.append(orm.Node, edge_filters=filters_forwards, with_incoming='sources')
+        rule_outgoing = UpdateRule(query_outgoing, max_iterations=1, track_edges=get_links)
+        rules += [rule_outgoing]
+
+    if links_forward and links_backward:
+        rules += [RuleSetWalkers(stash)]
+
+    if links_backward:
+        query_incoming = orm.QueryBuilder()
+        query_incoming.append(orm.Node, tag='sources')
+        query_incoming.append(orm.Node, edge_filters=filters_backwards, with_outgoing='sources')
+        rule_incoming = UpdateRule(query_incoming, max_iterations=1, track_edges=get_links)
+        rules += [rule_incoming]
+
+    rulesequence = RuleSequence(rules, max_iterations=max_iterations)
+
+    results = rulesequence.run(basket)
+
+    output = {}
+    output['nodes'] = results.nodes.keyset
+    output['links'] = None
+    if get_links:
```

Comment:
What is the point of the `get_links` argument if the return type is always going to be the same and the same computational cost is required, regardless of its value? If you always have to calculation but just put it as `None` in a dictionary, then why not just remove this argument. The caller can then just ignore it if they are not interested

Reply:
I'm not sure that the cost is the same. Indeed, the `get_links` keyword is passed to the Rules into `track_edges`, which is eventually passed to the querybuilder call as more projections to be made. I'm not sufficiently familiar with the QB to know how expensive are these extra projections though.

Reply:
OK let's keep this for now then